rust/src/csv/reader.rs

```diff
@@ -54,6 +57,120 @@ use crate::record_batch::RecordBatch;
 
 use self::csv_crate::{StringRecord, StringRecordsIntoIter};
 
+lazy_static! {
+    static ref DECIMAL_RE: Regex = Regex::new(r"^-?(\d+\.\d+)$").unwrap();
+    static ref INTEGER_RE: Regex = Regex::new(r"^-?(\d*.)$").unwrap();
+    static ref BOOLEAN_RE: Regex = RegexBuilder::new(r"^(true)$|^(false)$")
+        .case_insensitive(true)
+        .build()
+        .unwrap();
+}
+
+/// A sufficiently large constant for the maximum number of CSV records to read
+const MAX_INFERENCE_RECORDS: usize = 1_000_000;
+
+/// Infer the data type of a record
+fn infer_field_schema(string: &str) -> DataType {
+    // when quoting is enabled in the reader, these quotes aren't escaped, we default to Utf8 for them
+    if string.starts_with("\"") {
+        return DataType::Utf8;
+    }
+    // match regex in a particular order
+    if BOOLEAN_RE.is_match(string) {
+        return DataType::Boolean;
+    } else if DECIMAL_RE.is_match(string) {
+        return DataType::Float64;
+    } else if INTEGER_RE.is_match(string) {
+        return DataType::Int64;
+    } else {
+        return DataType::Utf8;
+    }
+}
+
+/// Infer the schema of a CSV file by reading through the first n records of the file,
+/// with `max_read_records` controlling the maximum number of records to read.
+/// 
+/// If `max_read_records` is not set, the current default `MAX_INFERENCE_RECORDS`
+/// is used.
+fn infer_file_schema(mut file: File, delimiter: u8, max_read_records: Option<usize>) -> Schema {
+
+    let mut csv_reader = csv::ReaderBuilder::new()
+        .delimiter(delimiter)
+        .from_reader(BufReader::new(file.try_clone().unwrap()));
+
+    // get header names
+    let headers = &csv_reader.headers().unwrap().clone();
+
+    // save the csv reader position after reading headers
+    let position = csv_reader.position().clone();
+
+    let my_headers: Vec<&str> = headers.iter().collect();
```

Comment:
nit: can we change `my_headers` to `headers` and move it below the above `let headers = &csv_reader..`?